### PR TITLE
transform.py and load.py tests

### PR DESCRIPTION
**Pull request by emily-curtis**
This branch handles the following things:

- Creates a test file `test_transform.py` based on the transform script
- Combines all tests into one script `test_etl.py` which tests all functionality for the scripts in the ETL pipeline

### CODE REVIEW

File: pipeline/test_transform.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    mock_get.return_value = mock_response
+    assert get_release_date_from_url("//example.com") == ""
+    mock_get.side_effect = Exception("Network Error")
+    assert get_release_date_from_url("//example.com") == ""
+
+def test_create_sales_dataframe():
+    """Test for create_sales_dataframe function."""
+    mock_sales_info = [
+        {
+            "item_type": "a",
```
Comment (ebradley12) [inline pipeline/test_transform.py:67]: Make sure this now reflects the updated transform script which converts 'a' to album and 't' to track.

File: pipeline/test_transform.py [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+def test_convert_date_format():
+    """Test for convert_date_format function."""
+    assert convert_date_format("1 October 2021") == "01-10-2021"
+    assert convert_date_format("invalid date") == "None"
+
+def test_convert_to_full_url():
+    """Test for convert_to_full_url function."""
+    assert convert_to_full_url("http://example.com") == "http://example.com"
+    assert convert_to_full_url("//example.com") == "https://example.com"
+
```
Comment (ebradley12) [inline pipeline/test_transform.py:30]: Really good test coverage

**Review by ebradley12 — COMMENTED** [review]
Really nice work. Now that the transform script is complete we can finish off this test suite :)

File: pipeline/test_transform.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            "item_type": "a",
+            "item_description": "Sample Album",
+            "album_title": "Mock Album",
+            "artist_name": "Mock Artist",
+            "country": "US",
+            "amount_paid_usd": 10.0,
+            "genres": ["Rock", "Pop"],
+            "release_date": "01-10-2021",
+            "sale_date": "2021-10-01"
+        }
```
Comment (GabrielNsiah) [inline pipeline/test_transform.py:76]:
```suggestion
        {
            "release_type": "a",
            "release_name": "Sample Album",
            "album_title": "Mock Album",
            "artist_name": "Mock Artist",
            "country": "US",
            "amount_paid_usd": 10.0,
            "genres": ["Rock", "Pop"],
            "release_date": "01-10-2021",
            "sale_date": "2021-10-01"
        }
```

If you're just testing functionality of `create_sales_dataframe` this is mostly fine, though I've changed around som names for the sake of clarity. Due to there being some extra cleaning/extensions now, it may be necessary to test that too.

File: pipeline/test_etl.py [hunk 4]
```diff
@@ -0,0 +1,232 @@
+"""test_etl.py: Unit tests for the ETL process.
+
+This script tests the entire ETL process:
+- Extraction of sales data from the API.
+- Transformation of extracted data.
+- Loading of transformed data into a database.
```
Comment (ebradley12) [inline pipeline/test_etl.py:6]: I like how the tests are in one script under classes

File: pipeline/test_etl.py [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+    @patch("extract.requests.get")
+    def test_extract_get_sales_information_failure(self, mock_get):
+        """Verify API failure handling."""
+        mock_response = MagicMock(status_code=404)
+        mock_get.return_value = mock_response
+
+        result = get_sales_information()
+        assert result == {}
+        mock_get.assert_called_once()
+
```
Comment (ebradley12) [inline pipeline/test_etl.py:55]: Maybe we could add a few more tests for extract to handle edge cases e.g. test_extract_get_sales_information_timeout() , test_extract_get_sales_information_connection_error()

File: pipeline/test_etl.py [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+                "release_date": "01-10-2021",
+                "sale_date": "2021-10-01",
+            }
+        ]
+        df = create_sales_dataframe(mock_sales_info)
+        assert isinstance(df, pd.DataFrame)
+        assert len(df) == 1
+        assert "release_type" in df.columns
+        assert df.iloc[0]["release_type"] == "t"
+
```
Comment (ebradley12) [inline pipeline/test_etl.py:166]: There seems to be 2 tests repeated here?

File: pipeline/test_etl.py [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
```
Comment (ebradley12) [inline pipeline/test_etl.py:167]: Could we add a test to check abbreviations 'a' and 't' are replaced by 'album' and 'track'

File: pipeline/test_etl.py [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+class TestLoad:
```
Comment (ebradley12) [inline pipeline/test_etl.py:168]: Great - these tests cover loads of different scenarios

File: pipeline/test_etl.py [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+        """Test insertion of a new artist."""
+        mock_cursor.fetchone.return_value = None
+        insert_artist("NewArtist", mock_cursor)
+        assert mock_cursor.execute.call_count == 2
+
+    def test_load_insert_genres_new(self, mock_cursor):
+        """Test insertion of a new genre."""
+        mock_cursor.fetchone.side_effect = [None, (1,)]
+        genre_id = insert_genres("NewGenre", mock_cursor)
+        assert genre_id == 1
```
Comment (ebradley12) [inline pipeline/test_etl.py:232]: Could we add a test for get_id_from_table() to ensure correct ID's are retrieved or returns None

File: pipeline/test_etl.py [hunk 10]
```diff
@@ -0,0 +1,232 @@
...
+        ]
+        df = create_sales_dataframe(mock_sales_info)
+        assert isinstance(df, pd.DataFrame)
+        assert len(df) == 1
+        assert "release_type" in df.columns
+        assert df.iloc[0]["release_type"] == "t"
+
+
+class TestLoad:
+    """Tests for the load phase of the ETL process."""
```
Comment (ebradley12) [inline pipeline/test_etl.py:169]: Could we add a test for the main_load() function

**Review by ebradley12 — APPROVED** [review]
Great work

**Review by lukieh2014 — APPROVED** [review]
great tests, and nice error handling added!
